neqo-transport/src/connection/mod.rs

```diff
@@ -1048,10 +1062,135 @@ impl Connection {
+                qinfo!(
+                    [self],
+                    "Received valid Initial packet with scid {:?} dcid {:?}",
+                    packet.scid(),
+                    packet.dcid()
+                );
+                self.set_state(State::WaitInitial);
+                self.loss_recovery.start_pacer(now);
+                self.crypto
+                    .states
+                    .init(self.quic_version, self.role, &packet.dcid());
+
+                // We need to make sure that we set this transport parameter.
+                // This has to happen prior to processing the packet so that
+                // the TLS handshake has all it needs.
+                if !self.retry_sent() {
+                    self.tps.borrow_mut().local.set_bytes(
+                        tparams::ORIGINAL_DESTINATION_CONNECTION_ID,
+                        packet.dcid().to_vec(),
+                    )
+                }
+            }
+            (PacketType::VersionNegotiation, State::WaitInitial, Role::Client) => {
+                match packet.supported_versions() {
+                    Ok(versions) => {
+                        if versions.is_empty() || versions.contains(&self.quic_version.as_u32()) {
+                            // Ignore VersionNegotiation packets that contain the current version.
+                            self.stats.borrow_mut().pkt_dropped("Invalid VN");
+                            return Ok(PreprocessResult::End);
+                        }
+
+                        self.set_state(State::Closed(ConnectionError::Transport(
+                            Error::VersionNegotiation,
+                        )));
+                        return Err(Error::VersionNegotiation);
+                    }
+                    Err(_) => {
+                        self.stats.borrow_mut().pkt_dropped("Invalid VN");
+                        return Ok(PreprocessResult::End);
+                    }
+                }
+            }
+            (PacketType::Retry, State::WaitInitial, Role::Client) => {
+                self.handle_retry(packet)?;
+                return Ok(PreprocessResult::Next);
+            }
+            (PacketType::Handshake, State::WaitInitial, Role::Client)
+            | (PacketType::Short, State::WaitInitial, Role::Client) => {
+                // This packet can't be processed now, but it could be a sign
+                // that Initial packets were lost.
+                // Resend Initial CRYPTO frames immediately a few times just
+                // in case.  As we don't have an RTT estimate yet, this helps
+                // when there is a short RTT and losses.
+                if first
+                    && self.is_valid_cid(packet.dcid())
+                    && self.stats.borrow().saved_datagrams <= EXTRA_INITIALS
+                {
+                    self.crypto.resend_unacked(PNSpace::Initial);
+                }
+            }
+            (PacketType::VersionNegotiation, ..)
```

Comment:
This case is a little whacky.  I was tempted to break it out specially, because you really want to stop processing packets if you get a Version Negotiation packet.  But those consume the entire packet and it should be dropped, so this is benign.